src/psyclone/domain/lfric/lfric_extract_driver_creator.py

```diff
@@ -915,12 +865,40 @@ def create(self, nodes, input_list, output_list,
 
         return file_container
 
+    # -------------------------------------------------------------------------
+    @staticmethod
+    def collect_all_required_modules(file_container):
+        '''Collects recursively all modules used in the file container.
+        It returns a dictionary, with the keys being all the (directly or
+        indirectly used modules)
+
+        :param file_container:
+        :type file_container: \
+            :py:class:`psyclone.psyir.psyir.nodes.FileContainer`
+
+        :returns: a dictionary, with the required module names as key, and \
+            as value a set of all modules required by the key module.
+        :rtype: Set[str]
+
+        '''
+        sym_tab = file_container.children[0].symbol_table
+
+        # Add all modules imported, except intrinsic ones
+        all_mods = set()
+        for symbol in sym_tab.symbols:
+            if isinstance(symbol, ContainerSymbol) and \
+                    ",intrinsic" not in symbol.name:
+                all_mods.add(symbol.name)
+
+        mod_manager = ModuleManager.get()
+        return mod_manager.get_all_dependencies_recursively(all_mods)
```

Comment:
I'm wondering whether this functionality should be part of fparser - it's something we will need to do when attempting to disambiguate symbols. I guess that's just something to bear in mind for the future though.

Reply:
I will say immediately *yes* ... aeh ... *no* ... aeh *maybe* to that :)
OK, that caused me some headache. I remembered that fparser had a cache, e.g.:
```
def get_module_file(name, directory, _cache={}):
```
but while it caches mapping of modules to file names, the actual parsed objects (or source code) are not cached at all as far as I could see. So from that point of view it would be nice to have in fparser ("yes").

On the other hand, I will need (e.g. in #1990) many PSyclone specific functions in there (e.g. needing the PSyIR, and using VariablesAccesses information) --> "no".

We could try adding the manager to fparser, and supporting a derived class to be used in PSyclone (which will use more powerful ModuleInfo objects to store psyir, external symbols, ...), But that would then imply a somewhat strong coupling between fparser's ModuleInfo and PSyclone's --> "maybe"

My suggestion: I add this to #2011 as an option to investigate - one concern I have is that it will take quite a bit of time (since I am not that familiar with fparser), and I would really prefer to focus on showing that driver extraction (mostly ... ignoring procedure pointers :)  ) works :)

Additionally, if the ModuleManager is part of fparser, it will do a LOT more parsing - I debugged it, when working on gungho, it would very frequently try to find infrastructure modules, which are all not found. If they would be found, this would likely trigger additional verification steps, which would also mean it might get quite a bit slower: depending on how exactly fparser triggers reading of imported modules ... I __think__ it will automatically do this by default for anything parser, and that could mean it might parse a LOT of additional files once fparser starts to find them, since they would cascade down: parsing a kernel will trigger parsing of constants_mod, logging_mod, kernel_mod, which all in turn might trigger more - we might end up parsing quite a bit of all infrastructure files anytime we run PSyclone.

Reply:
The more I look at it, (the fact that we deal with strings rather than PSyIR) the more I think it should be in fparser. It's not something that fparser would necessarily use by default - just functionality that could be used if requested.
I could try adding it to fparser once this PR is done. That way you can carry on pushing towards PSyKE :-)

Reply:
I've created https://github.com/stfc/fparser/issues/402 to look at implementing this in fparser.

Reply:
OK, I am happy to see if we can move part of this functionality to fparser.